File: core/src/test/java/io/typst/command/CommandTest.java
package io.typst.command;


import io.typst.command.algebra.Either;
import org.junit.jupiter.api.Test;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static io.typst.command.Command.pair;
import static io.typst.command.StandardArguments.intArg;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class CommandTest {
    private static final Argument<Integer> intTabArg = intArg.withTabCompletes(() -> Arrays.asList("10", "20"));
    private static final Argument<Integer> intTabArg2 = intArg.withTabCompletes(() -> Arrays.asList("30", "40"));
    // MyCommand = AddItem | RemoveItem | ...
    private static final Command.Mapping<MyCommand> itemCommand;
    private static final Command.Parser<MyCommand> pageCmd = Command.argument(PageItem::new, intTabArg, intTabArg2);

    static {
        itemCommand = Command.mapping(
                pair("open", Command.present(new OpenItemList()).withDescription("아이템 목록을 엽니다.")),
                // intArg: Argument<Integer>
                // strArg: Argument<String>
                pair("add", Command.argument(AddItem::new, intArg, StandardArguments.strArg)),
                pair("remove", Command.argument(RemoveItem::new, intArg)),
                pair("page", pageCmd),
                pair("lazy", Command.present(new AddItem(0, null))),
                pair("camelPage", pageCmd)
        );
    }

    private static final Command.Mapping<MyCommand> itemCommandWithFallback =
            itemCommand.withFallback(Command.present(new FallbackItem()));
    public static final Command.Mapping<MyCommand> itemCommandWithFallback2 =
            itemCommandWithFallback.withFallback(Command.argument(FallbackArg::new, intArg));
    public static final Command.Mapping<MyCommand> itemCommandWithFallback3 =
            itemCommandWithFallback.withFallback(Command.argument(FallbackOptArg::new, intArg.asOptional()));
    private static final Command<MyCommand> reloadCommand =
            Command.<MyCommand>present(new ReloadCommand()).withDescription("리로드합니다.");
    private static final Command<MyCommand> rootCommand =
            Command.mapping(
                    pair("item", itemCommand),
                    pair("reload", reloadCommand)
            );

    public interface MyCommand {
    }

    public static class AddItem implements MyCommand {
        private final int index;
        private final String name;

        public AddItem(int index, String name) {
            this.index = index;
            this.name = name;
        }

        @Override
        public String toString() {
            return "AddItem{" +
                    "index=" + index +
                    ", name='" + name + '\'' +
                    '}';
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            AddItem addItem = (AddItem) o;
            return index == addItem.index && Objects.equals(name, addItem.name);
        }

        @Override
        public int hashCode() {
            return Objects.hash(index, name);
        }
    }

    public static class RemoveItem implements MyCommand {
        public final Number index;

        public RemoveItem(Number index) {
            this.index = index;
        }
    }

    public static class PageItem implements MyCommand {
        public final int index;
        public final int indexB;

        public PageItem(int index, int indexB) {
            this.index = index;
            this.indexB = indexB;
        }
    }

    public static class OpenItemList implements MyCommand {
        @Override
        public boolean equals(Object obj) {
            return obj instanceof OpenItemList;
        }
    }

    public static class FallbackItem implements MyCommand {
        @Override
        public boolean equals(Object obj) {
            return obj instanceof FallbackItem;
        }
    }

    public static class FallbackArg implements MyCommand {
        public final int value;

        public FallbackArg(int value) {
            this.value = value;
        }

        @Override
        public boolean equals(Object o) {
            return o instanceof FallbackArg;
        }
    }

    public static class FallbackOptArg implements MyCommand {
        public final Integer value;

        @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
        public FallbackOptArg(Optional<Integer> value) {
            this.value = value.orElse(null);
        }

        @Override
        public boolean equals(Object o) {
            return o instanceof FallbackOptArg;
        }
    }

    public static class ReloadCommand implements MyCommand {
    }

    // executor
    @Test
    public void unit() {
        String[] args = new String[0];
        assertEquals(
                new Either.Left<>(new CommandFailure.FewArguments<>(args, 0, rootCommand)),
                Command.parse(args, rootCommand)
        );
    }

    @Test
    public void sub() {
        String[] args = new String[]{"item"};
        assertEquals(
                new Either.Left<>(new CommandFailure.FewArguments<>(args, 1, itemCommand)),
                Command.parse(args, rootCommand)
        );
    }

    @Test
    public void subUnknown() {
        String[] args = new String[]{"item", "unknownCommand"};
        assertEquals(
                new Either.Left<>(new CommandFailure.UnknownSubCommand<>(args, 1, itemCommand)),
                Command.parse(args, rootCommand)
        );
    }

    @Test
    public void present() {
        String[] args = new String[]{"item", "open"};
        assertEquals(
                new Either.Right<>(new CommandSuccess<>(args, 2, new OpenItemList(), itemCommand.getCommandMap().get("open"))),
                Command.parse(args, rootCommand)
        );
    }

    @Test
    public void argument() {
        int index = 0;
        String name = "someName";
        String[] args = new String[]{"item", "add", String.valueOf(index), name};
        assertEquals(
                new Either.Right<>(new CommandSuccess<>(args, args.length, new AddItem(index, name), itemCommand.getCommandMap().get("add"))),
                Command.parse(args, rootCommand)
        );
    }

    @Test
    public void help() {
        String[] args = new String[]{"item", "a"};
        Either<CommandFailure<MyCommand>, CommandSuccess<MyCommand>> result = Command.parse(args, rootCommand);
        CommandFailure<MyCommand> failure = result instanceof Either.Left
                ? ((Either.Left<CommandFailure<MyCommand>, CommandSuccess<MyCommand>>) result).getLeft()
                : null;
        if (failure instanceof CommandFailure.FewArguments) {
            CommandFailure.FewArguments<MyCommand> fewArgs = (CommandFailure.FewArguments<MyCommand>) failure;
            helpCommand(fewArgs.getArguments(), fewArgs.getIndex(), fewArgs.getCommand());
        } else if (failure instanceof CommandFailure.UnknownSubCommand) {
            CommandFailure.UnknownSubCommand<MyCommand> unknown = (CommandFailure.UnknownSubCommand<MyCommand>) failure;
            helpCommand(unknown.getArguments(), unknown.getIndex(), unknown.getCommand());
        }
    }

    private <A> void helpCommand(String[] args, int position, Command<A> cmd) {
        String[] succArgs = args.length >= 1
                ? Arrays.copyOfRange(args, 0, position)
                : new String[0];
        for (Map.Entry<List<String>, Command<A>> pair : Command.getEntries(cmd)) {
            List<String> usageArgs = pair.getKey().stream()
                    .flatMap(s -> Stream.concat(
                            Arrays.stream(succArgs),
                            Stream.of(s)
                    ))
                    .collect(Collectors.toList());
            String description = CommandSpec.from(pair.getValue()).getDescription();
            String suffix = description.isEmpty()
                    ? ""
                    : " - " + description;
            System.out.println("/" + String.join(" ", usageArgs) + suffix);
        }
    }

    @Test
    public void fallback() {
        String[] args = new String[]{"item"};
        Command.Mapping<MyCommand> commandMap = Command.mapping(
                pair("item", itemCommandWithFallback)
        );
        Either<CommandFailure<MyCommand>, CommandSuccess<MyCommand>> result = Command.parse(args, commandMap);
        assertEquals(
                new Either.Right<>(new CommandSuccess<>(args, 1, new FallbackItem(), itemCommandWithFallback.getFallback().orElse(null))),
                result
        );
    }

    @Test
    public void fallbackNone() {
        String[] args2 = new String[]{"item", "help"};
        Command.Mapping<MyCommand> commandMap = Command.mapping(
                pair("item", itemCommandWithFallback)
        );
        Either<CommandFailure<MyCommand>, CommandSuccess<MyCommand>> result2 = Command.parse(args2, commandMap);
        assertEquals(
                new Either.Left<>(new CommandFailure.UnknownSubCommand<>(args2, 1, itemCommandWithFallback)),
                result2
        );
    }

    @Test
    public void fallbackArg() {
        String[] args = new String[]{"item", "1"};
        Command.Mapping<MyCommand> commandMap = Command.mapping(
                pair("item", itemCommandWithFallback2)
        );
        Either<CommandFailure<MyCommand>, CommandSuccess<MyCommand>> result3 = Command.parse(args, commandMap);
        assertEquals(
                new Either.Right<>(new CommandSuccess<>(args, 2, new FallbackArg(1), itemCommandWithFallback2.getFallback().orElse(null))),
                result3
        );
    }

    @Test
    public void fallbackOptionalArg() {
        String[] args = new String[]{"item", "help"};
        Command.Mapping<MyCommand> commandMap = Command.mapping(
                pair("item", itemCommandWithFallback3)
        );
        Either<CommandFailure<MyCommand>, CommandSuccess<MyCommand>> result3 = Command.parse(args, commandMap);
        assertEquals(
                new Either.Left<>(new CommandFailure.UnknownSubCommand<>(args, 1, itemCommandWithFallback3)),
                result3
        );
    }

    // tab
    @Test
    public void tabUnit() {
        String[] args = new String[0];
        assertEquals(
                new CommandTabResult.Suggestions<>(Arrays.asList(
                        pair("item", Optional.of(itemCommand)),
                        pair("reload", Optional.of(reloadCommand))
                )),
                Command.tabComplete(args, rootCommand)
        );
    }

    @Test
    public void tabSub() {
        String[] args = new String[]{"item", ""};
        assertEquals(
                new CommandTabResult.Suggestions<>(Stream.of(
                        "open", "add", "remove", "page", "lazy", "camelPage"
                ).map(it ->
                        pair(it, Optional.ofNullable(itemCommand.getCommandMap().get(it)))
                ).collect(Collectors.toList())),
                Command.tabComplete(args, rootCommand)
        );
    }

    @Test
    public void tabSubUnknown() {
        String[] args = new String[]{"item", "unknownCommand"};
        assertEquals(
                new CommandTabResult.Suggestions<>(Collections.emptyList()),
                Command.tabComplete(args, rootCommand)
        );
    }

    @Test
    public void tabCustom() {
        assertEquals(
                new CommandTabResult.Suggestions<>(Stream.of("10", "20").map(
                        it -> pair(it, Optional.<Command<MyCommand>>of(pageCmd))
                ).collect(Collectors.toList())),
                Command.tabComplete(new String[]{"item", "page", ""}, rootCommand)
        );
        assertEquals(
                new CommandTabResult.Suggestions<>(Stream.of("10").map(
                        it -> pair(it, Optional.<Command<MyCommand>>of(pageCmd))
                ).collect(Collectors.toList())),
                Command.tabComplete(new String[]{"item", "page", "1"}, rootCommand)
        );
    }

    @Test
    public void tabCustom2() {
        assertEquals(
                new CommandTabResult.Suggestions<>(Stream.of("30", "40").map(
                        it -> pair(it, Optional.<Command<MyCommand>>of(pageCmd))
                ).collect(Collectors.toList())),
                Command.tabComplete(new String[]{"item", "page", "10", ""}, rootCommand)
        );
        assertEquals(
                new CommandTabResult.Suggestions<>(Stream.of("30").map(
                        it -> pair(it, Optional.<Command<MyCommand>>of(pageCmd))
                ).collect(Collectors.toList())),
                Command.tabComplete(new String[]{"item", "page", "10", "3"}, rootCommand)
        );
    }

    @Test
    public void tabUpperCase() {
        assertEquals(
                new CommandTabResult.Suggestions<>(Stream.of("30", "40").map(
                        it -> pair(it, Optional.<Command<MyCommand>>of(pageCmd))
                ).collect(Collectors.toList())),
                Command.tabComplete(new String[]{"item", "camelPage", "10", ""}, rootCommand)
        );
    }
}

File: core/src/test/kotlin/io/typst/command/CommandTest.kt
package io.typst.command

import io.typst.command.Command.pair
import io.typst.command.StandardArguments.intArg
import io.typst.command.StandardArguments.strArg
import io.typst.command.algebra.Either
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.util.*

class CommandTest {

    sealed interface MyCommand

    data class AddItem(val index: Int, val name: String?) : MyCommand
    data class RemoveItem(val index: Number) : MyCommand
    data class PageItem(val index: Int, val indexB: Int) : MyCommand
    data object OpenItemList : MyCommand
    data object FallbackItem : MyCommand
    data class FallbackArg(val value: Int) : MyCommand
    data class FallbackOptArg(val value: Int?) : MyCommand {
        constructor(value: Optional<Int>) : this(value.orElse(null))
    }
    data object ReloadCommand : MyCommand

    companion object {
        private val intTabArg = intArg.withTabCompletes { listOf("10", "20") }
        private val intTabArg2 = intArg.withTabCompletes { listOf("30", "40") }

        val pageCmd: Command.Parser<MyCommand> = Command.argument(::PageItem, intTabArg, intTabArg2)

        val itemCommand: Command.Mapping<MyCommand> = Command.mapping(
            pair("open", Command.present<MyCommand>(OpenItemList).withDescription("Opens the item list.")),
            pair("add", Command.argument(::AddItem, intArg, strArg)),
            pair("remove", Command.argument(::RemoveItem, intArg)),
            pair("page", pageCmd),
            pair("lazy", Command.present(AddItem(0, null))),
            pair("camelPage", pageCmd)
        )

        private val itemCommandWithFallback: Command.Mapping<MyCommand> =
            itemCommand.withFallback(Command.present(FallbackItem))

        val itemCommandWithFallback2: Command.Mapping<MyCommand> =
            itemCommandWithFallback.withFallback(Command.argument(::FallbackArg, intArg))

        val itemCommandWithFallback3: Command.Mapping<MyCommand> =
            itemCommandWithFallback.withFallback(Command.argument(::FallbackOptArg, intArg.asOptional()))

        private val reloadCommand: Command<MyCommand> =
            Command.present<MyCommand>(ReloadCommand).withDescription("Reloads the configuration.")

        val rootCommand: Command<MyCommand> = Command.mapping(
            pair("item", itemCommand),
            pair("reload", reloadCommand)
        )

        private fun left(failure: CommandFailure<MyCommand>): Either<CommandFailure<MyCommand>, CommandSuccess<MyCommand>> =
            Either.Left(failure)

        private fun right(success: CommandSuccess<MyCommand>): Either<CommandFailure<MyCommand>, CommandSuccess<MyCommand>> =
            Either.Right(success)
    }

    @Test
    fun `return FewArguments when no args provided`() {
        val args = emptyArray<String>()
        val result = Command.parse(args, rootCommand)

        assertThat(result).isEqualTo(left(CommandFailure.FewArguments(args, 0, rootCommand)))
    }

    @Test
    fun `return FewArguments for incomplete subcommand`() {
        val args = arrayOf("item")
        val result = Command.parse(args, rootCommand)

        assertThat(result).isEqualTo(left(CommandFailure.FewArguments(args, 1, itemCommand)))
    }

    @Test
    fun `return UnknownSubCommand for invalid subcommand`() {
        val args = arrayOf("item", "unknownCommand")
        val result = Command.parse(args, rootCommand)

        assertThat(result).isEqualTo(left(CommandFailure.UnknownSubCommand(args, 1, itemCommand)))
    }

    @Test
    fun `parse present command successfully`() {
        val args = arrayOf("item", "open")
        val result = Command.parse(args, rootCommand)

        assertThat(result).isEqualTo(
            right(CommandSuccess(args, 2, OpenItemList, itemCommand.commandMap["open"]))
        )
    }

    @Test
    fun `parse command with arguments`() {
        val index = 0
        val name = "someName"
        val args = arrayOf("item", "add", index.toString(), name)
        val result = Command.parse(args, rootCommand)

        assertThat(result).isEqualTo(
            right(CommandSuccess(args, args.size, AddItem(index, name), itemCommand.commandMap["add"]))
        )
    }

    @Test
    fun `use fallback when no subcommand matches`() {
        val args = arrayOf("item")
        val commandMap: Command.Mapping<MyCommand> = Command.mapping(
            pair("item", itemCommandWithFallback)
        )
        val result = Command.parse(args, commandMap)

        assertThat(result).isEqualTo(
            right(CommandSuccess(args, 1, FallbackItem, itemCommandWithFallback.fallback.orElse(null)))
        )
    }

    @Test
    fun `return UnknownSubCommand when fallback exists but input is invalid`() {
        val args = arrayOf("item", "help")
        val commandMap: Command.Mapping<MyCommand> = Command.mapping(
            pair("item", itemCommandWithFallback)
        )
        val result = Command.parse(args, commandMap)

        assertThat(result).isEqualTo(left(CommandFailure.UnknownSubCommand(args, 1, itemCommandWithFallback)))
    }

    @Test
    fun `parse fallback with argument`() {
        val args = arrayOf("item", "1")
        val commandMap: Command.Mapping<MyCommand> = Command.mapping(
            pair("item", itemCommandWithFallback2)
        )
        val result = Command.parse(args, commandMap)

        assertThat(result).isEqualTo(
            right(CommandSuccess(args, 2, FallbackArg(1), itemCommandWithFallback2.fallback.orElse(null)))
        )
    }

    @Test
    fun `parse fallback with optional argument`() {
        val args = arrayOf("item", "1")
        val commandMap: Command.Mapping<MyCommand> = Command.mapping(
            pair("item", itemCommandWithFallback3)
        )
        val result = Command.parse(args, commandMap)

        assertThat(result).isEqualTo(
            right(CommandSuccess(args, 2, FallbackOptArg(Optional.of(1)), itemCommandWithFallback3.fallback.orElse(null)))
        )
    }

    @Test
    fun `return UnknownSubCommand for invalid input with optional fallback`() {
        val args = arrayOf("item", "help")
        val commandMap: Command.Mapping<MyCommand> = Command.mapping(
            pair("item", itemCommandWithFallback3)
        )
        val result = Command.parse(args, commandMap)

        assertThat(result).isEqualTo(left(CommandFailure.UnknownSubCommand(args, 1, itemCommandWithFallback3)))
    }

    @Test
    fun `tab complete at root level`() {
        val args = emptyArray<String>()
        val result = Command.tabComplete(args, rootCommand)

        assertThat(result).isEqualTo(
            CommandTabResult.Suggestions(
                listOf(
                    pair("item", Optional.of(itemCommand)),
                    pair("reload", Optional.of(reloadCommand))
                )
            )
        )
    }

    @Test
    fun `tab complete subcommands`() {
        val args = arrayOf("item", "")
        val result = Command.tabComplete(args, rootCommand)

        assertThat(result).isEqualTo(
            CommandTabResult.Suggestions(
                listOf("open", "add", "remove", "page", "lazy", "camelPage").map {
                    pair(it, Optional.ofNullable(itemCommand.commandMap[it]))
                }
            )
        )
    }

    @Test
    fun `tab complete returns empty for unknown subcommand`() {
        val args = arrayOf("item", "unknownCommand")
        val result = Command.tabComplete(args, rootCommand)

        assertThat(result).isEqualTo(CommandTabResult.Suggestions<MyCommand>(emptyList()))
    }

    @Test
    fun `tab complete custom suggestions for first argument`() {
        assertThat(Command.tabComplete(arrayOf("item", "page", ""), rootCommand))
            .isEqualTo(
                CommandTabResult.Suggestions(
                    listOf("10", "20").map { pair(it, Optional.of<Command<MyCommand>>(pageCmd)) }
                )
            )

        assertThat(Command.tabComplete(arrayOf("item", "page", "1"), rootCommand))
            .isEqualTo(
                CommandTabResult.Suggestions(
                    listOf("10").map { pair(it, Optional.of<Command<MyCommand>>(pageCmd)) }
                )
            )
    }

    @Test
    fun `tab complete custom suggestions for second argument`() {
        assertThat(Command.tabComplete(arrayOf("item", "page", "10", ""), rootCommand))
            .isEqualTo(
                CommandTabResult.Suggestions(
                    listOf("30", "40").map { pair(it, Optional.of<Command<MyCommand>>(pageCmd)) }
                )
            )

        assertThat(Command.tabComplete(arrayOf("item", "page", "10", "3"), rootCommand))
            .isEqualTo(
                CommandTabResult.Suggestions(
                    listOf("30").map { pair(it, Optional.of<Command<MyCommand>>(pageCmd)) }
                )
            )
    }

    @Test
    fun `tab complete works with camelCase commands`() {
        assertThat(Command.tabComplete(arrayOf("item", "camelPage", "10", ""), rootCommand))
            .isEqualTo(
                CommandTabResult.Suggestions(
                    listOf("30", "40").map { pair(it, Optional.of<Command<MyCommand>>(pageCmd)) }
                )
            )
    }

    @Test
    fun `generate help from command failure`() {
        val args = arrayOf("item", "a")
        val result = Command.parse(args, rootCommand)

        val failure = (result as Either.Left).left
        assertThat(failure).isInstanceOf(CommandFailure.UnknownSubCommand::class.java)

        val unknown = failure as CommandFailure.UnknownSubCommand
        val entries = Command.getEntries(unknown.command)

        assertThat(entries).isNotEmpty
        assertThat(entries.map { it.key }).contains(listOf("open"))
    }
}